external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaTokenUtil.scala

```diff
@@ -223,14 +236,27 @@ private[spark] object KafkaTokenUtil extends Logging {
     }
   }
 
-  def isTokenAvailable(): Boolean = {
-    UserGroupInformation.getCurrentUser().getCredentials.getToken(
-      KafkaTokenUtil.TOKEN_SERVICE) != null
+  def findMatchingToken(
+      sparkConf: SparkConf,
+      bootStrapServers: String): Option[KafkaTokenClusterConf] = {
+    val tokens = UserGroupInformation.getCurrentUser().getCredentials.getAllTokens.asScala
+    val clusterConfigs = tokens
+      .filter(_.getService().toString().startsWith(TOKEN_SERVICE_PREFIX))
+      .map { token =>
+        KafkaTokenSparkConf.getClusterConfig(sparkConf, getClusterIdentifier(token.getService()))
+      }
+      .filter { clusterConfig =>
+        val pattern = Pattern.compile(clusterConfig.targetServersRegex)
+        Utils.stringToSeq(bootStrapServers).exists(pattern.matcher(_).matches())
```

Comment:
Also, an idea: wouldn't it be good to match the servers against the "auth" list too? That way the regex doesn't need to match the servers listed in the "auth" config, and in fact can be optional in case the "auth" list has all the servers you want.

Reply:
> That way the regex doesn't need to match the servers listed in the "auth" config

Not sure if I understand what you mean. `findMatchingToken` does the following:
* Get all available tokens
* Drops all non-Kafka tokens
* Gets a valid Kafka cluster config for each token (tokens are obtained based on this config)
* Does pattern matching agains the source/sink `bootstrap.servers`

This code part is independent from "auth" list.

You mean 2 step matching? Like first try to match against "auth" list and if not found try to match with regex?

Reply:
> Like first try to match against "auth" list and if not found try to match with regex?

Yes. That way for simple / small clusters you don't even need to define the regex, if all the servers are listed in the auth list.

Reply:
The most simple single cluster scenario works without any regex config by the user.

In the multi-cluster case even if providing a proper regex is an extra step I would keep the implementation as it is. As a user I would personally burn more time on understanding such fallback scenarios than writing a regex. If hostname pattern changes somehow then it has to be re-understand again (in general not big fan of fallbacks until it's absolutely not required).

If you think we can adapt the code because both solutions are good.

Reply:
Sounded like a pretty easy thing to implement, but anyway, not worth spending too much time on.